### PR TITLE
fixbug: RFC 135

### PR DESCRIPTION
**Pull request by iorisa**
feat: +SummarizeCode
feat: Move the logic related to project_name to WritePRD.
feat: Add unit tests
feat: The log for incrementally generated code can be referred to in the appendix section of [RFC 135](https://deepwisdom.feishu.cn/wiki/B1bcwZRdPiC8OYkweZYc9qWinSb).

### CODE REVIEW

File: metagpt/roles/engineer.py [hunk 1]
```diff
@@ -105,43 +118,87 @@ async def _act(self) -> Message | None:
...
+            summary = await todo.run()
+            summary_filename = Path(todo.context.design_filename).with_suffix(".md").name
+            dependencies = {todo.context.design_filename, todo.context.task_filename}
+            for filename in todo.context.codes_filenames:
+                rpath = src_relative_path / filename
+                dependencies.add(str(rpath))
+            await code_summaries_pdf_file_repo.save(filename=summary_filename, content=summary,
+                                                    dependencies=dependencies)
+            is_pass, reason = await self._is_pass(summary)
+            if not is_pass:
```
Comment (iorisa) [inline metagpt/roles/engineer.py:151]: fix me

File: metagpt/actions/design_api.py [hunk 2]
```diff
@@ -58,15 +59,15 @@
 """,
         "FORMAT_EXAMPLE": """
 [CONTENT]
-{
+{{
     "Implementation approach": "We will ...",
-    "project_name": "snake_game",
+    "project_name": "{project_name}",
```
Comment (better629) [inline metagpt/actions/design_api.py:64]: How about "Project name", keep the same.

File: metagpt/actions/write_prd.py [hunk 3]
```diff
@@ -53,7 +55,7 @@
 {{
     "Language": "", # str, use the same language as the user requirement. en_us / zh_cn etc.
     "Original Requirements": "", # str, place the polished complete original requirements here
-    "project_name": "", # str, name it like game_2048 / web_2048 / simple_crm etc.
+    "project_name": "{project_name}", # str, if it's empty, name it with snake case style, like game_2048 / web_2048 / simple_crm etc.
```
Comment (better629) [inline metagpt/actions/write_prd.py:58]: Project Name. same as below

File: metagpt/roles/engineer.py [hunk 4]
```diff
@@ -216,16 +273,16 @@ async def _new_code_actions(self):
...
         # Generate a SummarizeCode action for each pair of (system_design_doc, task_doc).
         summarizations = {}
-        for filename in changed_src_files:
-            dependencies = src_file_repo.get_dependency(filename=filename)
+        for filename in src_files:
+            dependencies = await src_file_repo.get_dependency(filename=filename)
             ctx = CodeSummarizeContext.loads(filenames=dependencies)
             if ctx not in summarizations:
-                summarizations[ctx] = set()
+                summarizations[ctx] = []
```
Comment (better629) [inline metagpt/roles/engineer.py:283]: use  summarizations=defaultdict(list) to simplify the code

File: metagpt/actions/write_prd.py (hunk 3 above)
Comment (iorisa) [reply metagpt/actions/write_prd.py]: fixed

File: metagpt/actions/design_api.py (hunk 2 above)
Comment (iorisa) [reply metagpt/actions/design_api.py]: done

**Comment by stellaHSR** [comment]
LGTM

**Pull request merged**